### PR TITLE
Support doc filtering in ateles

### PR DESCRIPTION
**Pull request by garrensmith**
Add the ability to filter docs for the changes feed using a javascript
filter or map function.

Requires this PR in CouchDB https://github.com/apache/couchdb/pull/3699

## To test

Apply this diff to CouchDB

```diff
diff --git a/rebar.config.script b/rebar.config.script
index 98fdcfdc7..c768cb5a9 100644
--- a/rebar.config.script
+++ b/rebar.config.script
@@ -159,7 +159,8 @@ DepDescs = [
 {meck,             "meck",             {tag, "0.9.2"}},
 {recon,            "recon",            {tag, "2.5.0"}},
 {passage,          "passage",          {tag, "CouchDB-0.2.6-1"}},
-{thrift_protocol,  "thrift-protocol",  {tag, "0.1.5"}}
+{thrift_protocol,  "thrift-protocol",  {tag, "0.1.5"}},
+{ateles, {url, "https://github.com/cloudant-labs/ateles"}, {branch, "filter-doc-support"}}
 ].
 
 WithProper = lists:keyfind(with_proper, 1, CouchConfig) == {with_proper, true}.
diff --git a/rel/overlay/etc/default.ini b/rel/overlay/etc/default.ini
index a65d667dc..5a549f154 100644
--- a/rel/overlay/etc/default.ini
+++ b/rel/overlay/etc/default.ini
@@ -343,7 +343,7 @@ bind_address = 127.0.0.1
 ; The list of modules that implement the couch_eval
 ; beahvior for executing provided code in design
 ; documents.
-;javascript = couch_js
+javascript = ateles
 ;query = mango_eval
 
 [mango]
diff --git a/rel/reltool.config b/rel/reltool.config
index a1cc938c1..67e087464 100644
--- a/rel/reltool.config
+++ b/rel/reltool.config
@@ -58,6 +58,8 @@
         passage,
         thrift_protocol,
         couch_prometheus,
+        gpb,
+        ateles,
         %% extra
         recon
     ]},
@@ -115,6 +117,8 @@
     {app, passage, [{incl_cond, include}]},
     {app, thrift_protocol, [{incl_cond, include}]},
     {app, couch_prometheus, [{incl_cond, include}]},
+    {app, gpb, [{incl_cond, include}]},
+    {app, ateles, [{incl_cond, include}]},
 
     %% extra
     {app, recon, [{incl_cond, include}]}
```

All tests should pass. Specially `changes_test.exs`. The filtering works as follows https://docs.couchdb.org/en/main/api/database/changes.html#filtering

### CODE REVIEW

File: src/ateles.erl [hunk 1]
```diff
@@ -123,14 +125,35 @@ validate_doc_update(DDoc, EditDoc, DiskDoc, UserCtx, SecObj) ->
     JsonEditDoc = couch_doc:to_json_obj(EditDoc, [revs]),
     JsonDDoc = couch_doc:to_json_obj(DDoc, []),
     JsonDiskDoc = couch_query_servers:json_doc(DiskDoc),
-    ddoc_prompt(JsonDDoc, <<"validate_doc_update">>, [JsonEditDoc, JsonDiskDoc, UserCtx, SecObj]).
+    ddoc_prompt(JsonDDoc, <<"validate">>, [<<"validate_doc_update">>], [JsonEditDoc, JsonDiskDoc, UserCtx, SecObj]).
 
-ddoc_prompt({Props} = DDoc, FunName, Args) ->
+filter_view(DDoc, VName, Docs) ->
+    JsonDocs = [couch_query_servers:json_doc(Doc) || Doc <- Docs],
+    JsonDDoc = couch_doc:to_json_obj(DDoc, []),
```
Comment (iilyak) [inline src/ateles.erl:133]: `couch_js` based version of the feature uses `Options` to  pass `revs_limit` [here](https://github.com/apache/couchdb/blob/main/src/couch_js/src/couch_js_query_servers.erl#L443:L444). Is it no longer  needed?

Comment (garrensmith) [reply src/ateles.erl:133]: I couldn't figure out why it would be needed. But thinking now it could be to limit the size of the document?

Comment (iilyak) [reply src/ateles.erl:133]: Here is  where  it  was  first   introduced https://github.com/apache/couchdb/commit/23e5f92a15cdc868622481d2ff63ca3770110fae

**Comment by iilyak** [comment]
Please update testing instructions to point to correct branch  name
```

+{ateles, {url, "https://github.com/cloudant-labs/ateles"}, {branch, "filter-replication-couch-js"}}
```
should be 
```

+{ateles, {url, "https://github.com/cloudant-labs/ateles"}, {branch, "filter-doc-support"}}
```

File: src/ateles.erl (hunk 1 above)
Comment (garrensmith) [reply src/ateles.erl:133]: I've added in the `Options` params.

**Comment by iilyak** [comment]
I don't know why but I cannot test it on my system.

```
diff --git a/rebar.config.script b/rebar.config.script
index 98fdcfdc7..0379ea806 100644
--- a/rebar.config.script
+++ b/rebar.config.script
@@ -159,7 +159,9 @@ DepDescs = [
 {meck,             "meck",             {tag, "0.9.2"}},
 {recon,            "recon",            {tag, "2.5.0"}},
 {passage,          "passage",          {tag, "CouchDB-0.2.6-1"}},
-{thrift_protocol,  "thrift-protocol",  {tag, "0.1.5"}}
+{thrift_protocol,  "thrift-protocol",  {tag, "0.1.5"}},
+{ateles, {url, "https://github.com/cloudant-labs/ateles"}, {branch, "filter-doc-support"}},
+{gpb, {url, "https://github.com/tomas-abrahamsson/gpb"}, {branch, "master"}}
 ].

 WithProper = lists:keyfind(with_proper, 1, CouchConfig) == {with_proper, true}.
diff --git a/rel/overlay/etc/default.ini b/rel/overlay/etc/default.ini
index a65d667dc..5a549f154 100644
--- a/rel/overlay/etc/default.ini
+++ b/rel/overlay/etc/default.ini
@@ -343,7 +343,7 @@ bind_address = 127.0.0.1
 ; The list of modules that implement the couch_eval
 ; beahvior for executing provided code in design
 ; documents.
-;javascript = couch_js
+javascript = ateles
 ;query = mango_eval

 [mango]
diff --git a/rel/reltool.config b/rel/reltool.config
index a1cc938c1..9295c8e3e 100644
--- a/rel/reltool.config
+++ b/rel/reltool.config
@@ -27,6 +27,7 @@
         xmerl,
         %% couchdb
         aegis,
+        ateles,
         b64url,
         bear,
         chttpd,
@@ -47,6 +48,7 @@
         ets_lru,
         fabric,
         folsom,
+        gpb,
         hyper,
         ibrowse,
         jaeger_passage,
@@ -84,6 +86,8 @@

     %% couchdb
     {app, aegis, [{incl_cond, include}]},
+    {app, gpb, [{incl_cond, include}]},
+    {app, ateles, [{incl_cond, include}]},
     {app, b64url, [{incl_cond, include}]},
     {app, bear, [{incl_cond, include}]},
     {app, chttpd, [{incl_cond, include}]},
```

```
ERROR: Conflicting dependencies for config: [{"From: ateles",
                                              {".*",
                                               {git,
                                                "https://github.com/apache/couchdb-config",
                                                {branch,"master"}}}},
                                             {"From: couchdb",
                                              {".*",
                                               {git,
                                                "https://github.com/apache/couchdb-config.git",
                                                {tag,"2.1.9"}}}}]
ERROR: Conflicting dependencies for jiffy: [{"From: couchdb",
                                             {".*",
                                              {git,
                                               "https://github.com/apache/couchdb-jiffy.git",
                                               {tag,"CouchDB-1.0.5-1"}}}},
                                            {"From: ateles",
                                             {".*",
                                              {git,
                                               "https://github.com/davisp/jiffy",
                                               {branch,"master"}}}}]
You have configured Apache CouchDB, time to relax. Relax.

❯ make eunit  apps=ateles
==> ateles (compile)
[ 34%] Built target boost
[100%] Built target ateles
ERROR: The gpb library is not present in code path!
ERROR: compile failed while processing /Users/iilyak@ca.ibm.com/dev/fdb-dbcore/src/couchdb/src/ateles: rebar_abort
make: *** [couch] Error 1

❯ ls src/gpb/
ls src/gpb/
COPYING.LIB          README.md
....

❯ ls src/gpb/ebin/

```

`gpb` is not compiled. I run `./configure` and I did `make`.

**Comment by iilyak** [comment]
I tried a fresh clone and have same result. 

```
Scanning dependencies of target ateles
[ 60%] Building CXX object CMakeFiles/ateles.dir/ateles.pb.cc.o
[ 69%] Building CXX object CMakeFiles/ateles.dir/c_src/js.cc.o
[ 69%] Building CXX object CMakeFiles/ateles.dir/c_src/connection.cc.o
[ 73%] Building CXX object CMakeFiles/ateles.dir/c_src/js_worker.cc.o
[ 78%] Building CXX object CMakeFiles/ateles.dir/c_src/main.cc.o
[ 82%] Building CXX object CMakeFiles/ateles.dir/c_src/message.cc.o
[ 86%] Building CXX object CMakeFiles/ateles.dir/c_src/server.cc.o
[ 91%] Building CXX object CMakeFiles/ateles.dir/c_src/stats.cc.o
[ 95%] Building CXX object CMakeFiles/ateles.dir/c_src/util.cc.o
[100%] Linking CXX executable ateles
[100%] Built target ateles
ERROR: The gpb library is not present in code path!
ERROR: compile failed while processing /Users/iilyak@ca.ibm.com/dev/couchdb-integrate-ateles/src/ateles: rebar_abort
make: *** [couch] Error 1
```

**Comment by iilyak** [comment]
The testing  instructions are not complete.  I'll log the set of  steps I  used for posterity.  

```
diff --git a/rebar.config.script b/rebar.config.script
index 4c3565e54..0379ea806 100644
--- a/rebar.config.script
+++ b/rebar.config.script
@@ -155,11 +155,13 @@ DepDescs = [
 {jaeger_passage,   "jaeger-passage",   {tag, "CouchDB-0.1.14-4"}},
 {jiffy,            "jiffy",            {tag, "CouchDB-1.0.5-1"}},
 {local,            "local",            {tag, "0.2.1"}},
-{mochiweb,         "mochiweb",         {tag, "CouchDB-v2.21.0-1"}},
+{mochiweb,         "mochiweb",         {tag, "v2.21.0"}},
 {meck,             "meck",             {tag, "0.9.2"}},
 {recon,            "recon",            {tag, "2.5.0"}},
 {passage,          "passage",          {tag, "CouchDB-0.2.6-1"}},
-{thrift_protocol,  "thrift-protocol",  {tag, "0.1.5"}}
+{thrift_protocol,  "thrift-protocol",  {tag, "0.1.5"}},
+{ateles, {url, "https://github.com/cloudant-labs/ateles"}, {branch, "filter-doc-support"}},
+{gpb, {url, "https://github.com/tomas-abrahamsson/gpb"}, {branch, "master"}}
```

```
diff --git a/rel/reltool.config b/rel/reltool.config
index a1cc938c1..9295c8e3e 100644
--- a/rel/reltool.config
+++ b/rel/reltool.config
@@ -27,6 +27,7 @@
         xmerl,
         %% couchdb
         aegis,
+        ateles,
         b64url,
         bear,
         chttpd,
@@ -47,6 +48,7 @@
         ets_lru,
         fabric,
         folsom,
+        gpb,
         hyper,
         ibrowse,
         jaeger_passage,
@@ -84,6 +86,8 @@

     %% couchdb
     {app, aegis, [{incl_cond, include}]},
+    {app, gpb, [{incl_cond, include}]},
+    {app, ateles, [{incl_cond, include}]},
```

 ```
./configure
rm src/ateles/rebar.config
make -C src/ateles init
make -C src/ateles/_build
mkdir -p src/ateles/priv/
cp src/ateles/_build/ateles src/ateles/priv/ateles
make
```

Start `ateles` in  separate tab `cd && src/ateles &&  ./priv/ateles -p 8444`
Run  test suite `make elixir-suite`

**Review by iilyak — APPROVED** [review]
+1

**Pull request merged**